### ISSUE TITLE
[Research] Federated Learning for Privacy-Preserving Multi-User EEG Model Training

### ISSUE THREAD

**Issue by LakshmiSravyaVedantham** [LakshmiSravyaVedantham/DreamAnalysis#44]
## Paper 1
**Title:** Fuzzy ensemble-based federated learning for EEG-based emotion recognition in Internet of Medical Things
**Year:** 2025
**Venue:** Journal of King Saud University (Elsevier)
**URL:** https://www.sciencedirect.com/science/article/abs/pii/S2452414X25000135

## Paper 2
**Title:** Privacy-Preserving EEG Data Generation: A Federated Split Learning
**Year:** 2025
**Venue:** SECRYPT 2025
**URL:** https://www.scitepress.org/Papers/2025/135291/135291.pdf

## Paper 3
**Title:** Privacy-preserving dementia classification from EEG via hybrid-fusion EEGNetv4 and federated learning
**Year:** 2025
**Venue:** Frontiers in Computational Neuroscience
**URL:** https://www.frontiersin.org/journals/computational-neuroscience/articles/10.3389/fncom.2025.1617883/full

## Key Contribution
The fuzzy ensemble FL framework combines TCN, LSTM, and GRU models with a Gompertz function-based fuzzy rank approach for combining predictions. Uses asynchronous dropout for efficient federated weight updates. Validated on GAMEEMO, SEED, and DEAP datasets with improved accuracy over baselines.

The SECRYPT 2025 paper introduces federated split learning with privacy-adaptive autoencoders — EEG data never leaves the device, only model gradients are shared. Critical because EEG signals are biometric identifiers that can reveal medical conditions.

## Relevance to NeuralDreamWorkshop
As NeuralDreamWorkshop scales to multiple users, per-user EEG data is sensitive biometric information. Federated learning enables:
1. Training a shared model from multiple users' data without centralizing raw EEG
2. Each user's data stays on their device / local session
3. Only model weight updates (gradients) are shared with the server
4. Collective learning improves accuracy for all users while preserving privacy

This connects directly to the "per-user personalization" roadmap item — FL provides the infrastructure for learning from all users to benefit each individual.

### Files to Create
- `ml/models/federated_trainer.py` — Federated learning coordinator:
  ```python
  class FederatedEEGTrainer:
      """Server-side federated learning coordinator."""
      def __init__(self, base_model, n_clients, aggregation='fedavg'):
          self.global_model = base_model
          self.client_updates = []

      def receive_update(self, client_id: str, model_delta: Dict):
          """Receive gradient update from a client."""

      def aggregate(self) -> Dict:
          """FedAvg: weighted average of client model updates."""

      def broadcast(self) -> Dict:
          """Send updated global model weights to all clients."""
  ```
- `ml/models/federated_client.py` — Client-side local training:
  ```python
  class FederatedEEGClient:
      """Client-side local training on user's EEG data."""
      def local_train(self, local_data, epochs=5):
          """Train on local data, return weight delta (not raw data)."""

      def apply_global_update(self, global_weights):
          """Update local model with aggregated global weights."""
  ```
- `ml/training/train_federated.py` — Simulation script to test FL with DEAP/SEED split across simulated clients

### Files to Modify
- `ml/api/routes.py`:
  - Add `POST /federated/submit-update` — Client submits local model delta
  - Add `GET /federated/global-model` — Client fetches latest global model
  - Add `GET /federated/status` — Federated training progress (rounds, participants, accuracy)
  - Add `POST /federated/opt-in` — User opts into federated learning
- `ml/models/emotion_classifier.py` — Add `get_model_delta()` and `apply_model_delta()` methods for FL compatibility
- `ml/models/online_learner.py` — Wire FL client into existing online learning pipeline
- `server/routes.ts` — Add privacy consent UI endpoints + user preference storage
- `shared/schema.ts` — Add `federated_consent` boolean to user settings table

### Privacy Architecture
```
User Device (Browser/App)
  └── Local EEG data (never leaves device)
  └── Local model training (5 epochs on new data)
  └── Compute weight delta = new_weights - old_weights
  └── POST /federated/submit-update (only delta, not data)

Server
  └── Collect deltas from N opted-in users
  └── FedAvg aggregation (weighted by sample count)
  └── Broadcast updated global model
  └── No raw EEG ever stored on server
```

### Differential Privacy (from SECRYPT paper)
```python
# Add noise to gradients before sending (local differential privacy)
def add_dp_noise(model_delta, epsilon=1.0):
    sigma = sensitivity / epsilon
    for key in model_delta:
        model_delta[key] += torch.normal(0, sigma, size=model_delta[key].shape)
    return model_delta
```

## Expected Impact
- **Privacy:** Raw EEG never leaves user's device
- **Accuracy:** Collective learning from N users improves model for all (+5-10%)
- **Personalization:** Each user gets global model + local fine-tuning = best of both worlds
- **Trust:** Users more willing to participate knowing data stays local
- **Regulatory:** GDPR/HIPAA compliant by design

## Complexity: HIGH
- Federated learning infrastructure requires careful synchronization
- Client-side training in browser requires ONNX/WASM or server-side with user data isolation
- Privacy guarantees need formal analysis (epsilon-delta DP)
- Testing FL with simulated clients before real deployment
- Database schema changes for consent tracking
